File: backend/services/transaction/src/transaction/models/transaction.py
import uuid
from datetime import date, datetime, timezone
from decimal import Decimal
from enum import StrEnum

from sqlalchemy import Date, DateTime, ForeignKey, Index, Numeric, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from transaction.database import Base


class TransactionType(StrEnum):
    EXPENSE = "expense"
    INCOME = "income"


class Category(StrEnum):
    """
    Fixed taxonomy — keeps AI analysis consistent.
    New categories require a migration + model update intentionally.
    """
    FOOD = "food"
    SHOPPING = "shopping"
    ENTERTAINMENT = "entertainment"
    TRANSPORT = "transport"
    HEALTH = "health"
    EDUCATION = "education"
    UTILITIES = "utilities"
    OTHER = "other"


class Transaction(Base):
    __tablename__ = "transactions"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    # Injected by Kong from JWT — this service never receives or validates JWTs directly
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False, index=True)

    amount: Mapped[Decimal] = mapped_column(
        Numeric(precision=15, scale=2), nullable=False
    )
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="VND")
    type: Mapped[TransactionType] = mapped_column(
        String(10), nullable=False, default=TransactionType.EXPENSE
    )
    category: Mapped[Category] = mapped_column(String(20), nullable=False)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Date of the transaction (user-supplied, can differ from created_at)
    transaction_date: Mapped[date] = mapped_column(Date, nullable=False)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    __table_args__ = (
        # Common query: user's transactions in a date range
        Index("ix_transactions_user_date", "user_id", "transaction_date"),
    )

File: frontend/services/transaction.ts
import { api } from "@/lib/api";

export type TransactionType = "expense" | "income";
export type Category =
  | "food"
  | "shopping"
  | "entertainment"
  | "transport"
  | "health"
  | "education"
  | "utilities"
  | "other";

export interface TransactionPayload {
  amount: number;
  currency?: string;
  type?: TransactionType;
  category: Category;
  note?: string;
  transaction_date: string; // ISO date string
}

export interface TransactionResponse {
  id: string;
  user_id: string;
  amount: number;
  currency: string;
  type: TransactionType;
  category: Category;
  note?: string;
  transaction_date: string;
  created_at: string;
}

export interface PaginatedTransactions {
  items: TransactionResponse[];
  total: number;
  page: number;
  page_size: number;
}

export const transactionApi = {
  list: (params?: { page?: number; page_size?: number; date_from?: string; date_to?: string }) =>
    api.get<PaginatedTransactions>("/transactions", { params }).then((r) => r.data),

  get: (id: string) =>
    api.get<TransactionResponse>(`/transactions/${id}`).then((r) => r.data),

  create: (data: TransactionPayload) =>
    api.post<TransactionResponse>("/transactions", data).then((r) => r.data),

  update: (id: string, data: Partial<TransactionPayload>) =>
    api.patch<TransactionResponse>(`/transactions/${id}`, data).then((r) => r.data),

  delete: (id: string) =>
    api.delete(`/transactions/${id}`),
};
